src/Microsoft.AspNetCore.Mvc.Core/Internal/AssemblyPartDiscoveryProvider.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    internal static class AssemblyPartDiscoveryProvider
+    {
+        public static void ConfigureApplicationParts(ApplicationPartManager partManager, string entryPointAssemblyName)
+        {
+            var entryAssembly = Assembly.Load(new AssemblyName(entryPointAssemblyName));
+            var entryModel = AssemblyPartDiscoveryModel.ResolveEntryPoint(entryAssembly);
+
+            foreach (var part in GetPartsInOrder(entryModel))
+            {
+                var operation = ConfigureApplicationPartManager.GetConfigureOperation(part);
+                operation.Configure(partManager, part);
+            }
+        }
+
+        // Internal for unit testing.
+        internal static IEnumerable<AssemblyPartDiscoveryModel> GetPartsInOrder(AssemblyPartDiscoveryModel entryModel)
+        {
+            // Consider
+            // App -> [ AdditionalParts: App.Views ] 
+            // App.deps.json
+            //  - Component1
+            //  - ClassLib11 -> [ AdditionalParts: ModuleComponent, Component1 ]
+            //  - ClassLib2
+            //  - ModuleComponent -> [ AdditionalPart: ThemingModule ]
+
+            // We want to ensure that parts are processed in the following order:
+            //  - App, App.Views, ClassLib1, Component1, ModuleComponent, ThemingModule, ClassLib2
+
+            // Here are the rules for ordering:
+            // * Peers are ordered by name
+            // * AdditionalParts appear immediately after the parent part.
+            return EnumerateModel(entryModel).Distinct();
```

Comment:
The order in which parts get to configure. In the default case, this determines the order of the parts